src/NetherPortalScanner.cpp

```diff
@@ -0,0 +1,291 @@
+				}
+			}
+		}
+	}
+}
+
+
+
+
+
+bool cNetherPortalScanner::IsValidBuildLocation(Vector3i a_BlockPos)
+{
+	// Check the base
+	for (int i = 0; i < SearchSolidBaseWidth; i++)
+	{
+		for (int j = 0; j < PortalLength; j++)
+		{
+			BLOCKTYPE blocktype = m_World->GetBlock(a_BlockPos.x + i, a_BlockPos.y, a_BlockPos.z + j);
+			if (!cBlockInfo::IsSolid(blocktype))
+			{
+				return false;
+			}
+
+			// Check the airspace
+			for (int k = 1; k < PortalHeight; k++)
+			{
+				blocktype = m_World->GetBlock(a_BlockPos.x + i, a_BlockPos.y + k, a_BlockPos.z + j);
+				if (blocktype != E_BLOCK_AIR)
+				{
+					return false;
+				}
+			}
+		}
+	}
+	return true;
+}
+
+
+
+
+
+bool cNetherPortalScanner::OnAllChunksAvailable(void)
+{
+	if (m_FoundPortal)
+	{
+		// Find the bottom of this portal
+		while (m_World->GetBlock(m_PortalLoc.x, m_PortalLoc.y, m_PortalLoc.z) == E_BLOCK_NETHER_PORTAL)
+		{
+			m_PortalLoc.y -= 1.0;
+		}
+		m_PortalLoc.y += 1.0;
+
+		// Figure out which way the portal is facing
+		int BXP = m_World->GetBlock(m_PortalLoc.x + 1, m_PortalLoc.y, m_PortalLoc.z);
+		int BXM = m_World->GetBlock(m_PortalLoc.x - 1, m_PortalLoc.y, m_PortalLoc.z);
+		if ((BXP == E_BLOCK_NETHER_PORTAL) || (BXM == E_BLOCK_NETHER_PORTAL))
+		{
+			// The long axis is along X
+			m_Dir = 0;
+		}
+		else
+		{
+			// The long axis is along Z
+			m_Dir = 1;
+		}
+	}
+	else
+	{
+		// Scan the area for a suitable location
+		int minx = FloorC(m_Position.x) - BuildSearchRadius;
+		int minz = FloorC(m_Position.z) - BuildSearchRadius;
+		int maxx = FloorC(m_Position.x) + BuildSearchRadius;
+		int maxz = FloorC(m_Position.z) + BuildSearchRadius;
+		int maxy = m_MaxY;
+		std::vector<Vector3i> Possibilities;
+		int x, y, z;
+		for (y = 0; y < maxy - PortalHeight; y++)
+		{
+			for (x = minx; x < maxx - PortalLength; x++)
+			{
+				for (z = minz; z < maxz - SearchSolidBaseWidth; z++)
+				{
+					Vector3i Location = Vector3i(x, y, z);
+					if (IsValidBuildLocation(Location))
+					{
+						Possibilities.push_back(Vector3i(x, y, z));
+					}
+				}
+			}
+		}
+
+		if (Possibilities.size() > 0)
+		{
+			m_BuildPlatform = false;
+
+			// Find the nearest
+			double DistanceToClosest = (Possibilities[0] - m_Position).SqrLength();
+			Vector3i Closest = Possibilities[0];
+			for (const auto & itr : Possibilities)
+			{
+				double Distance = (itr - m_Position).SqrLength();
+				if (Distance < DistanceToClosest)
+				{
+					DistanceToClosest = Distance;
+					Closest = itr;
+				}
+			}
+
+			m_PortalLoc = Closest;
+		}
+	}
+	return true;
+}
+
+
+
+
+
+void cNetherPortalScanner::BuildNetherPortal(Vector3i a_Location, int a_Direction, bool a_IncludePlatform)
+{
+	int x = a_Location.x;
+	int y = a_Location.y;
+	int z = a_Location.z;
+
+	// Clear a 3x4x4 area starting right above the base
+	for (int i = 0; i < SearchSolidBaseWidth; i++)
+	{
+		for (int j = 0; j < PortalLength; j++)
+		{
+			for (int k = 1; k < PortalHeight; k++)
+			{
+				if (a_Direction)
```

Comment:
Don't use an `int` like a `bool`. Ints should be compared to explicit values, bools should be used directly.
`if (intValue == 4)` vs `if (boolValue)`